pandas-stubs/_typing.pyi

```diff
@@ -192,11 +263,10 @@ AstypeArg: TypeAlias = (
     | TimedeltaDtypeArg
     | TimestampDtypeArg
     | CategoryDtypeArg
-    | ExtensionDtype
+    | DtypeObj
     | type[object]
+    | str
```

Comment:
I don't think we should accept a general string here.  The whole idea of the other aliases is to constrain which strings are acceptable.

Reply:
`str` must be acceptable, as e.g. a string might be provided dynamically. Same is if you accept generic dtype object, which one must if `s.astye(s.dtype)` should work without raising a warning.

The correct way to handle this is to be careful with overload order, going from special to generic.
All of the existing tests still work with this change.

Reply:
It makes the type check too wide.  It allows `pd.DataFrame().astype("foobar")` to pass.  I'd rather have the stubs catch the most used cases (not dynamic strings).  Having said that, if we want `s.astype(s.dtype)` to work, maybe we should just say that `s.dtype()` returns `AsTypeArg`

Reply:
If there was a way to specify a non-literal string, that would be ideal.  If it sees a literal, the literal should be verified. I am not aware of a way to achieve that with current typing features.

Reply:
Another note - I think in subjective cases like this, you have to ask, what is the most common use case. I would contend dynamic/runtime based `astype` calls are atypical. The majority of the time, it's a static transformation, and the developer is aware of the source dtype and target dtype. In that case, they would hard-code the destination type.

Reply:
@Dr-Irv `pd.DataFrame().astype("foobar")` is a `ValueError`, type hints are for avoiding `TypeError`.

Reply:
But I get that catching typos would be very useful.

Reply:
@gandhis1 As an example for dynamic astype(string): I have some data-pipelines for reading csv-files where I store config files with the target schema (e.g. column dtypes).

Reply:
@gandhis1 @Dr-Irv In python 3.11 [`LiteralString`](https://docs.python.org/3/library/typing.html#typing.LiteralString) was added, I wonder id this can be used to raise a typing error if none of the defined literals is matched? This would give best of both worlds.

Reply:
I asked this here: https://github.com/python/typing/discussions/1434

Reply:
Hm, doesn't seem possible currently. Possibly in the future if both `Intersection` and `Not` are supported: https://github.com/python/typing/issues/801#issuecomment-1646171898

Reply:
> @gandhis1 As an example for dynamic astype(string): I have some data-pipelines for reading csv-files where I store config files with the target schema (e.g. column dtypes).

That's an interesting example, but that's a case where the argument you are using is dynamic, and you're trying to check it in a static typing context.  If you config files had an incorrect string `"foobar"`not covered by the stubs, your code would fail at runtime.  

IMHO, the goal of the static type checks is to catch things before runtime.  So we have two alternatives:

1. Do not include `str` (as it is right now in the stubs) as a valid argument.  Disadvantage is that you can't use a dynamic string as an argument as in your use case.  Advantage is it covers only the valid strings that are statically declared.
2. Include `str` as a valid argument.  Disadvantage is that we don't help people who use `.astype("ing")` instead of `.astype("int")` prior to run time.  Advantage is your use case.

I would argue that people use `.astype("some-string")` more than have a dynamic string as the argument.  So that's why I would prefer (1) above.